eth_abi/registry.py

```diff
@@ -0,0 +1,100 @@
+from .utils.parsing import collapse_type
+
+
+class ABIRegistryError(Exception):
+    pass
+
+
+class NoEntryFound(ABIRegistryError):
+    pass
+
+
+class PredicateMapping:
+    def __init__(self):
+        self._values = {}
+        self._labeled_predicates = {}
+
+    def add(self, predicate, value, label=None):
+        if predicate in self._values:
+            raise ValueError('Predicate already exists')
+
+        if label is not None:
+            if label in self._labeled_predicates:
+                raise ValueError('Predicate with label "{}" already exists'.format(label))
+
+            self._labeled_predicates[label] = predicate
+
+        self._values[predicate] = value
+
+    def remove_by_predicate(self, predicate):
+        # Delete the predicate mapping to the previously stored value
+        try:
+            del self._values[predicate]
+        except KeyError:
+            raise KeyError('Predicate not found')
+
+        # Delete any label which refers to this predicate
+        try:
+            label = self._label_for_predicate(predicate)
+        except ValueError:
+            pass
+        else:
+            del self._labeled_predicates[label]
+
+    def _label_for_predicate(self, predicate):
+        for key, value in self._labeled_predicates.items():
+            if value is predicate:
+                return key
+
+        raise ValueError('Predicate not referred to by any label')
+
+    def remove_by_label(self, label):
+        try:
+            predicate = self._labeled_predicates[label]
+        except KeyError:
+            raise KeyError('Label not found')
+
+        del self._labeled_predicates[label]
+        del self._values[predicate]
+
+
+class ABIRegistry:
+    def __init__(self):
+        self._encoders = PredicateMapping()
+
+    def register_encoder(self, lookup, encoder, label=None):
+        """
+        Registers an encoder under the given type lookup and optional label.
+        """
+        if callable(lookup):
+            self._encoders.add(lookup, encoder, label)
+            return
+
+        if isinstance(lookup, str):
+            def predicate(base, sub, arrlist):
```

Comment:
Extracting this to the module level and currying it should make this a bit cleaner.

```python
from cytoolz import curry

@curry
def static_predicate(lookup, base, sub, arrlist):
    return collapse_type(base, sub, arrlist) == lookup

... # then using it later.
self.encoders.add(static_predicate(lookup), encoder, lookup)
```

Reply:
I guess that means we'd need to add cytoolz under `install_requires` in `setup.py`.  Are there any problems that might come up as a result of both eth-abi and eth-utils depending on cytoolz (e.g. different version specs for cytoolz package might become out of sync, etc.)?  I guess I've worked on other projects where this might have come up but never thought of this before.

Reply:
`cytoolz` is pretty standard across our codebases as well as being quite stable.  Maybe worth doing a quick check that the version ranges intersect with whatever is specified in `eth-utils`